### PR TITLE
ddl: constant or timezone-dependent expressions in partition should not be allowed

### PR DESCRIPTION
**Pull request by qiffang**
### What problem does this PR solve? 
Mysql do not allow Constant, random or timezone-dependent expressions in partitioning function.
But TiDB do not check whether the field is legal 
If TiDB do not check the field, there will be the same bug - https://bugs.mysql.com/bug.php?id=50036 (MySQL has fixed long before)

```
create table t1(time_recorded timestamp) PARTITION BY RANGE ( TO_DAYS(time_recorded) ) (PARTITION P0 VALUES LESS THAN (1559192604))

create table t2(id int) partition by range(DAY(id)) (partition p1 values less than (10));

create table t3(id int) partition by range(4) (partition p1 values less than (10));

create table t4(id int) partition by hash(4) partitions 4;
....
``` 
The above statements can not be executed in Mysql because of **Constant, random or timezone-dependent expressions in (sub)partitioning function are not allowed**, but they can be executed in TiDB.

### What is changed and how it works?
Translate MySQL code to TiDB
https://github.com/mysql/mysql-server/blob/5.7/sql/sql_partition.cc#L1072
https://github.com/mysql/mysql-server/blob/5.7/sql/item_timefunc.h#L343
https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L445

### Check List 

Tests 

 - Unit test
   partition_test.go which is existed works well
 - Manual test (add detailed scripts or steps below)
  TiDB has the same action with MySQL.When I execute the above statements, they can not be executed because of Constant, random or timezone-dependent expressions in (sub)partitioning function are not allowed

### CODE REVIEW

**Comment by shenli** [comment]
@qiffang Please edit the PR description, remove the items that are not applicable in the check list.

**Comment by qiffang** [comment]
> @qiffang Please edit the PR description, remove the items that are not applicable in the check list.

Done

**Comment by coocood** [comment]
The old unit test doesn't cover the changed logic.

**Comment by qiffang** [comment]
> The old unit test doesn't cover the changed logic.

Add new cases for cost and timezone-dependent type

File: ddl/partition.go [hunk 1]
```diff
@@ -163,32 +163,114 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+// The only exception is UNIX_TIMESTAMP() which returns the internal
+// representation of a TIMESTAMP argument verbatim, and thus does not depend on
+// the timezone.
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L445
+func defaultTimezoneDependent(columns []*expression.Column) bool {
+	return !hasTimestampField(columns)
+}
+
+var (
+	funcCallExprMap = map[string]isTimezoneDependentProcessor{
```
Comment (coocood) [inline ddl/partition.go]: The name `funcCallExprMap` is too generic, we can't infer what it does by the name.

**Comment by coocood** [comment]
@qiffang 
Have you considered using a single function `isTimezoneDependent` and `switch` instead of `map`?
It will be easier to see if the code is covered. and we can just test the function in the unit test, it's easier to find the code uncovered by the unit test.

**Comment by qiffang** [comment]
> @qiffang
> Have you considered using a single function `isTimezoneDependent` and `switch` instead of `map`?
> It will be easier to see if the code is covered. and we can just test the function in the unit test, it's easier to find the code uncovered by the unit test.

I do not think switch cases is a better way for several reasons
1) If we use `map`, then find a funcName only cost O(1), but `switch/case` need O(n) 
2) There are several kinds, so the code looks more redundant

**Comment by coocood** [comment]
Switch a constant value is not O(n), the performance is similar to map.
And for each case, there would be only 1 more line compares to the map.

**Comment by qiffang** [comment]
> Switch a constant value is not O(n), the performance is similar to map.
> And for each case, there would be only 1 more line compares to the map.

If we use `switch case`, it will like this
```
switch v.FnName.L {
		case ast.Day, ast.DayOfMonth, ast.DayOfWeek, ast.DayOfYear,
			ast.Hour, ast.MicroSecond, ast.Minute, ast.Mod, ast.Month, ast.Quarter, ast.Second, ast.TimeToSec, ast.ToDays, ast.ToSeconds, ast.Weekday, ast.Year, ast.YearWeek:
			xxx
                 case ast.Abs, ast.Ceiling, ast.DateDiff, ast.Extract, ast.Floor, ast.Mod:
                         xxx
		case ast.UnixTimestamp:
                         xxxx
``` 
If the FnName is `ast.Year`,  it is similar to map?

**Comment by coocood** [comment]
@qiffang 
This way, the code is much less than map, performance is similar, of course.

**Comment by qiffang** [comment]
> @qiffang
> This way, the code is much less than map, performance is similar, of course.

I do not think it is similar,  of course, if we think there is only 10+ const in one case, there is almost no differences.
But if the function is called very frequency, and there are many const in one case line, they are different.
So i do not like this kind of writing even though this function is called low frequency.

**Comment by coocood** [comment]
@qiffang 
Please do some benchmark to verify your assumption.

**Comment by qiffang** [comment]
> @qiffang
> Please do some benchmark to verify your assumption.

For checking the performance, we need to consider cpu branch predictability

Fortunately, there have been a benchmark tests - https://github.com/jackc/go_map_vs_switch

With the const increasing, the map is better.

**Comment by coocood** [comment]
The result shows that the performance is similar, obvious not O(N) as you thought.

**Comment by qiffang** [comment]
> The result shows that the performance is similar, obvious not O(N) as you thought.

The Conclusion of tests
```
Conclusion
The original question I wanted to answer was if there would be a performance impact in converting a switch statement to a map lookup. I assumed that switches would be faster. I was wrong. Maps are usually faster, often several times faster.

Does this mean we should prefer using maps over switches? No. While the percentage changes are substantial, the absolute time difference is small. It will only matter if the branch is executed millions of times per second and the actual work performed is minimal. Additionally, even if that is the case, the memory access pattern and branch prediction success rate have a bigger impact on performance than the choice of a switch or map.
```

File: ddl/db_partition_test.go [hunk 2]
```diff
@@ -160,14 +160,15 @@ func (s *testIntegrationSuite9) TestCreateTableWithPartition(c *C) {
...
-	assertErrorCode(c, tk, sql9, tmysql.ErrPartitionFunctionIsNotAllowed)
+	// this sql is not allowed in mysql because of the result is const
+	//sql9 := `create TABLE t9 (
+	//col1 int
+	//)
+	//partition by range( case when col1 > 0 then 10 else 20 end ) (
+	//	partition p0 values less than (2),
+	//	partition p1 values less than (6)
+	//);`
+	//assertErrorCode(c, tk, sql9, tmysql.ErrPartitionFunctionIsNotAllowed)
```
Comment (tiancaiamao) [inline ddl/db_partition_test.go]: Why comment out this test? The error is changed?

File: ddl/db_partition_test.go [hunk 3]
```diff
@@ -1449,3 +1450,54 @@ func (s *testIntegrationSuite4) TestPartitionErrorCode(c *C) {
...
+	tk.MustExec("use test_db_with_partition_const")
+
+	sql1 := `create table t1 ( id int ) 
+		partition by range(4) (
+		partition p1 values less than (10)
+		);`
+	assertErrorCode(c, tk, sql1, tmysql.ErrWrongExprInPartitionFunc)
+
+	sql2 := `create table t2 ( time_recorded timestamp ) 
+		partition by range(TO_DAYS(time_recorded)) (
```
Comment (tiancaiamao) [inline ddl/db_partition_test.go]: TO_DAYS function works on datetime type? 

How about adding this one:
```
create table t2 ( time_recorded datetime ) 
		partition by range(TO_DAYS(time_recorded)) ( 
...
```

File: ddl/db_partition_test.go [hunk 4]
```diff
@@ -1449,3 +1450,54 @@ func (s *testIntegrationSuite4) TestPartitionErrorCode(c *C) {
...
+		partition p1 values less than (1559192604)
+		);`
+	assertErrorCode(c, tk, sql2, tmysql.ErrWrongExprInPartitionFunc)
+
+	sql3 := `create table t3 ( id int ) 
+		partition by range(DAY(id)) (
```
Comment (tiancaiamao) [inline ddl/db_partition_test.go]: Is this a timezone dependent expression?

**Comment by tiancaiamao** [comment]
I've benchmarked using `switch` vs `array` for dispatching before: when there are many cases, `array` is much better than switch.

But I think it's `switch` vs `map` performance is not something matters here.
Let's do not digress. If @coocood worry about test coverage, we can 

```
for ... {
   "create table xx .. partition by hash"   + ast.String()
}
```

to cover all the isTimezoneDependent functions ... but it seems meaningless

File: ddl/partition.go [hunk 5]
```diff
@@ -163,32 +163,114 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+		}
+	}
+
+	return false
+}
+
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L412
+func hasTimeField(columns []*expression.Column) bool {
+	for _, c := range columns {
+		if c.GetType().Tp == mysql.TypeDatetime {
```
Comment (crazycs520) [inline ddl/partition.go]: `if c.GetType().Tp == mysql.TypeDatetime || c.GetType().Tp == mysql.TypeDuration {`
/* TypeDuration original name was TypeTime, renamed to TypeDuration to resolve the conflict with Go type Time.*/

File: ddl/db_partition_test.go (hunk 3 above)
Comment (qiffang) [reply ddl/db_partition_test.go]: Please check the error code.
It does not work on date time type, i will check it in next line
```
assertErrorCode(c, tk, sql2, tmysql.ErrWrongExprInPartitionFunc)
```

File: ddl/db_partition_test.go (hunk 4 above)
Comment (qiffang) [reply ddl/db_partition_test.go]: Please check mysql the document i posted
This sql is not allowed in mysql, the error check is in next line

File: ddl/partition.go (hunk 5 above)
Comment (qiffang) [reply ddl/partition.go]: Got it, thanks.

File: ddl/db_partition_test.go (hunk 4 above)
Comment (tiancaiamao) [reply ddl/db_partition_test.go]: OK... it just makes me confused that the value of  `DAY(integer)`  depends on the timezone

File: ddl/db_partition_test.go (hunk 2 above)
Comment (qiffang) [reply ddl/db_partition_test.go]: Updated, thanks

File: ddl/db_partition_test.go (hunk 3 above)
Comment (qiffang) [reply ddl/db_partition_test.go]: > I mean, how about adding this line:
> 
> ```
> tk.MustExec("create table t2 ( time_recorded datetime ) 
> 		partition by range(TO_DAYS(time_recorded)) ( 
> ...")
> ```
> 
> If my memory serve me right, there are some table partition tests use `to_days` as partition function
> https://github.com/pingcap/tidb-test/blob/master/mysql_test/t/partition_range.test#L17

This sql is right, the rules please check the description or mysql comments.
Thanks

Comment (qiffang) [reply ddl/db_partition_test.go]: > https://github.com/pingcap/tidb-test/blob/master/mysql_test/t/partition_range.test#L17

For `to_days` method, adding following cases
1. to_days(time)
2. to_days(timestamp)
3. to_days(datetime)
4. to_days(date)

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @coocood

**Comment by coocood** [comment]
> But do you think it is necessary to replace map with switch?
We are not replacing map with switch, switch is the original code.
This PR change switch to map, I think we should not make this change.

**Comment by qiffang** [comment]
> > But do you think it is necessary to replace map with switch?
> 
> @qiffang
> We are not replacing map with switch, switch is the original code.
> This PR change switch to map, I think we should not make this change.

As your suggestions, use `switch` to replace `map`

File: ddl/partition.go [hunk 6]
```diff
@@ -177,32 +177,94 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
 	return nil
 }
 
+// For partition tables, mysql do not support Constant, random or timezone-dependent expressions
+// Based on mysql code to check whether field is valid, every time related type has check_valid_arguments_processor function.
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_timefunc
+type isTimezoneDependentProcessor func(columns []*expression.Column) bool
```
Comment (coocood) [inline ddl/partition.go]: This type can be removed

Comment (qiffang) [reply ddl/partition.go]: Removed

**Comment by coocood** [comment]
LGTM

**Pull request merged**